File: docker/scripts/export_modelling.py
#!/usr/bin/env python3
"""
Export Modelling (MDL) from WrenAI

This script exports the complete modelling definition from WrenAI system:
- Models, columns, relationships, views
- Export to JSON file in docker/data/
"""

import requests
import json
import os
from datetime import datetime

# Configuration
UI_GRAPHQL_ENDPOINT = "http://wrenai-wren-ui-1:3000/api/graphql"  # Use container hostname
OUTPUT_DIR = "/app/data"  # Inside container path

def get_latest_deploy_hash():
    """Get the latest deployment hash from deploy logs"""
    
    query = """
    query {
        __typename
    }
    """
    
    # For now, we'll use a known hash or get it from deploy_log table
    # This would need to be implemented to query the database directly
    # or add a GraphQL query for deploy logs
    
    # Return the hash we used in our test
    return "852c9584f33bf6df0006a983b332dfa524282eeb"

def query_mdl(hash_value):
    """Query MDL via GraphQL API using deployment hash"""
    
    query = f"""
    query GetMDL {{
        getMDL(hash: "{hash_value}") {{
            hash
            mdl
        }}
    }}
    """
    
    payload = {"query": query}
    
    try:
        response = requests.post(
            UI_GRAPHQL_ENDPOINT,
            json=payload,
            headers={"Content-Type": "application/json"},
            timeout=30
        )
        
        if response.status_code == 200:
            result = response.json()
            if 'errors' in result:
                print(f"❌ GraphQL Error: {result['errors']}")
                return None
            else:
                return result['data']['getMDL']
        else:
            print(f"❌ HTTP Error: {response.status_code} - {response.text}")
            return None
            
    except Exception as e:
        print(f"❌ Request Error: {e}")
        return None

def decode_mdl(encoded_mdl):
    """Decode base64 encoded MDL to JSON"""
    import base64
    
    try:
        decoded_bytes = base64.b64decode(encoded_mdl)
        decoded_str = decoded_bytes.decode('utf-8')
        return json.loads(decoded_str)
    except Exception as e:
        print(f"❌ Error decoding MDL: {e}")
        return None

def export_mdl_to_json(mdl_data, output_path):
    """Export MDL to JSON file"""
    
    if not mdl_data:
        print("⚠️  No MDL data to export")
        return False
    
    try:
        with open(output_path, 'w', encoding='utf-8') as jsonfile:
            json.dump(mdl_data, jsonfile, indent=2, ensure_ascii=False)
        
        # Print summary
        models_count = len(mdl_data.get('models', []))
        relationships_count = len(mdl_data.get('relationships', []))
        views_count = len(mdl_data.get('views', []))
        
        print(f"✅ Successfully exported MDL to {output_path}")
        print(f"   📊 Models: {models_count}")
        print(f"   🔗 Relationships: {relationships_count}")
        print(f"   👁️  Views: {views_count}")
        print(f"   🗄️  DataSource: {mdl_data.get('dataSource', 'Unknown')}")
        print(f"   🏷️  Catalog: {mdl_data.get('catalog', 'Unknown')}")
        print(f"   📋 Schema: {mdl_data.get('schema', 'Unknown')}")
        
        return True
        
    except Exception as e:
        print(f"❌ Error exporting MDL: {e}")
        return False

def main():
    print('=' * 80)
    print('📤 WrenAI Modelling (MDL) Export Tool')
    print('=' * 80)
    print(f'⏰ Export time: {datetime.now().strftime("%Y-%m-%d %H:%M:%S")}')
    print()
    
    # Ensure output directory exists
    os.makedirs(OUTPUT_DIR, exist_ok=True)
    
    # Get latest deployment hash
    print("🔍 Getting latest deployment hash...")
    hash_value = get_latest_deploy_hash()
    print(f"   Hash: {hash_value}")
    print()
    
    # Export MDL
    print("📋 Exporting MDL...")
    mdl_result = query_mdl(hash_value)
    
    if mdl_result is not None:
        # Decode the base64 encoded MDL
        print("🔓 Decoding MDL...")
        mdl_data = decode_mdl(mdl_result['mdl'])
        
        if mdl_data:
            # Export to JSON file
            timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
            mdl_file = os.path.join(OUTPUT_DIR, f'modelling_export_{timestamp}.json')
            
            if export_mdl_to_json(mdl_data, mdl_file):
                print()
                print("✅ MDL export completed successfully!")
                print(f"📁 File saved: {mdl_file}")
            else:
                print("❌ Failed to export MDL")
                exit(1)
        else:
            print("❌ Failed to decode MDL")
            exit(1)
    else:
        print("❌ Failed to query MDL")
        exit(1)
    
    print()
    print("🎉 Export process completed!")

if __name__ == "__main__":
    main() 

File: docker/scripts/export_modelling.js
#!/usr/bin/env node
/**
 * Export Modelling (MDL) from WrenAI
 * 
 * This script exports the complete modelling definition from WrenAI system:
 * - Models, columns, relationships, views
 * - Export to JSON file in /app/data/
 */

const fs = require('fs');
const path = require('path');

// Configuration
const UI_GRAPHQL_ENDPOINT = "http://localhost:3000/api/graphql";
const OUTPUT_DIR = "/app/data";

// Use dynamic import for fetch
async function loadFetch() {
  const { default: fetch } = await import('node-fetch');
  return fetch;
}

function getCurrentTimestamp() {
  const now = new Date();
  return now.toISOString().replace(/[:.]/g, '-').slice(0, 19);
}

async function getLatestDeployHash() {
  // For now, return the known hash
  // This should be enhanced to query the database or add GraphQL query
  return "852c9584f33bf6df0006a983b332dfa524282eeb";
}

async function queryMDL(hashValue) {
  const fetch = await loadFetch();
  
  const query = `
    query GetMDL {
      getMDL(hash: "${hashValue}") {
        hash
        mdl
      }
    }
  `;
  
  const payload = {
    query: query
  };
  
  try {
    const response = await fetch(UI_GRAPHQL_ENDPOINT, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(payload)
    });
    
    if (response.ok) {
      const result = await response.json();
      if (result.errors) {
        console.log(`❌ GraphQL Error: ${JSON.stringify(result.errors)}`);
        return null;
      } else {
        return result.data.getMDL;
      }
    } else {
      console.log(`❌ HTTP Error: ${response.status} - ${await response.text()}`);
      return null;
    }
  } catch (error) {
    console.log(`❌ Request Error: ${error.message}`);
    return null;
  }
}

function decodeMDL(encodedMDL) {
  try {
    const decodedStr = Buffer.from(encodedMDL, 'base64').toString('utf-8');
    return JSON.parse(decodedStr);
  } catch (error) {
    console.log(`❌ Error decoding MDL: ${error.message}`);
    return null;
  }
}

function exportMDLToJSON(mdlData, outputPath) {
  if (!mdlData) {
    console.log("⚠️  No MDL data to export");
    return false;
  }
  
  try {
    fs.writeFileSync(outputPath, JSON.stringify(mdlData, null, 2), 'utf-8');
    
    // Print summary
    const modelsCount = mdlData.models ? mdlData.models.length : 0;
    const relationshipsCount = mdlData.relationships ? mdlData.relationships.length : 0;
    const viewsCount = mdlData.views ? mdlData.views.length : 0;
    
    console.log(`✅ Successfully exported MDL to ${outputPath}`);
    console.log(`   📊 Models: ${modelsCount}`);
    console.log(`   🔗 Relationships: ${relationshipsCount}`);
    console.log(`   👁️  Views: ${viewsCount}`);
    console.log(`   🗄️  DataSource: ${mdlData.dataSource || 'Unknown'}`);
    console.log(`   🏷️  Catalog: ${mdlData.catalog || 'Unknown'}`);
    console.log(`   📋 Schema: ${mdlData.schema || 'Unknown'}`);
    
    return true;
  } catch (error) {
    console.log(`❌ Error exporting MDL: ${error.message}`);
    return false;
  }
}

async function main() {
  console.log('=' .repeat(80));
  console.log('📤 WrenAI Modelling (MDL) Export Tool');
  console.log('=' .repeat(80));
  console.log(`⏰ Export time: ${new Date().toISOString()}`);
  console.log();
  
  // Ensure output directory exists
  if (!fs.existsSync(OUTPUT_DIR)) {
    fs.mkdirSync(OUTPUT_DIR, { recursive: true });
  }
  
  // Get latest deployment hash
  console.log("🔍 Getting latest deployment hash...");
  const hashValue = await getLatestDeployHash();
  console.log(`   Hash: ${hashValue}`);
  console.log();
  
  // Export MDL
  console.log("📋 Exporting MDL...");
  const mdlResult = await queryMDL(hashValue);
  
  if (mdlResult !== null) {
    // Decode the base64 encoded MDL
    console.log("🔓 Decoding MDL...");
    const mdlData = decodeMDL(mdlResult.mdl);
    
    if (mdlData) {
      // Export to JSON file
      const timestamp = getCurrentTimestamp();
      const mdlFile = path.join(OUTPUT_DIR, `modelling_export_${timestamp}.json`);
      
      if (exportMDLToJSON(mdlData, mdlFile)) {
        console.log();
        console.log("✅ MDL export completed successfully!");
        console.log(`📁 File saved: ${mdlFile}`);
      } else {
        console.log("❌ Failed to export MDL");
        process.exit(1);
      }
    } else {
      console.log("❌ Failed to decode MDL");
      process.exit(1);
    }
  } else {
    console.log("❌ Failed to query MDL");
    process.exit(1);
  }
  
  console.log();
  console.log("🎉 Export process completed!");
}

// Install node-fetch if not available
async function ensureDependencies() {
  try {
    require.resolve('node-fetch');
  } catch (e) {
    console.log("📦 Installing node-fetch...");
    const { execSync } = require('child_process');
    execSync('npm install node-fetch@2', { stdio: 'inherit' });
  }
}

if (require.main === module) {
  ensureDependencies().then(() => {
    main().catch(error => {
      console.error('❌ Unexpected error:', error);
      process.exit(1);
    });
  });
} 